lib/pacemaker/pcmk_output.c

```diff
@@ -1266,6 +1267,79 @@ node_action_xml(pcmk__output_t *out, va_list args)
     return pcmk_rc_ok;
 }
 
+PCMK__OUTPUT_ARGS("node-name", "uint32_t", "const char *")
+static int
+node_name_default(pcmk__output_t *out, va_list args)
+{
+    uint32_t node_id = va_arg(args, uint32_t);
+    const char *node_name = va_arg(args, const char *);
+
+    if (node_id == 0) {
+        return out->info(out, "Node name: %s (local node)",
+                         pcmk__s(node_name, "unknown"));
+    }
+    return out->info(out, "Node name: %s (nodeid=%" PRIu32 ")",
+                     pcmk__s(node_name, "unknown"), node_id);
+}
+
+PCMK__OUTPUT_ARGS("node-name", "uint32_t", "const char *")
+static int
+node_name_html(pcmk__output_t *out, va_list args)
+{
+    uint32_t node_id = va_arg(args, uint32_t);
+    const char *node_name = va_arg(args, const char *);
+
+    pcmk__output_create_xml_text_node(out, "span", "Node name: ");
+    pcmk__output_create_xml_text_node(out, "span",
+                                      pcmk__s(node_name, "unknown"));
+
+    if (node_id == 0) {
+        pcmk__output_create_xml_text_node(out, "span", " (local node)");
+    } else {
+        char *id_s = crm_strdup_printf("%" PRIu32, node_id);
+
+        pcmk__output_create_xml_text_node(out, "span", " (nodeid=");
+        pcmk__output_create_xml_text_node(out, "span", id_s);
+        pcmk__output_create_xml_text_node(out, "span", ")");
+        free(id_s);
+    }
+
+    return pcmk_rc_ok;
+}
+
+PCMK__OUTPUT_ARGS("node-name", "uint32_t", "const char *")
+static int
+node_name_text(pcmk__output_t *out, va_list args)
+{
+    if (!out->is_quiet(out)) {
+        return node_name_default(out, args);
+    } else {
+        const char *node_id G_GNUC_UNUSED = va_arg(args, const char *);
+        const char *node_name = va_arg(args, const char *);
+
+        pcmk__formatted_printf(out, "%s\n", pcmk__s(node_name, "unknown"));
+        return pcmk_rc_ok;
+    }
+}
+
+PCMK__OUTPUT_ARGS("node-name", "uint32_t", "const char *")
+static int
+node_name_xml(pcmk__output_t *out, va_list args)
+{
+    uint32_t node_id = va_arg(args, uint32_t);
+    const char *node_name = va_arg(args, const char *);
+
+    char *id_s = crm_strdup_printf("%" PRIu32, node_id);
+
+    // For XML, preserve id=0 instead of using "local node" as in other formats
+    pcmk__output_create_xml_node(out, "node-name",
+                                 XML_ATTR_ID, id_s,
+                                 XML_ATTR_NAME, pcmk__s(node_name, "unknown"),
```

Comment:
The expectation is to do the `out->err()` for the "can't find a node name" situation at the caller level; if someone forces a message call, then print "unknown"